src/sentry/digests/notifications.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import absolute_import
+
+import functools
+import itertools
+import logging
+from collections import (
+    OrderedDict,
+    defaultdict,
+    namedtuple,
+)
+
+from sentry.app import tsdb
+from sentry.digests import Record
+from sentry.models import (
+    Project,
+    Group,
+    GroupStatus,
+    Rule,
+)
+from sentry.utils.dates import to_timestamp
+
+
+logger = logging.getLogger('sentry.digests')
+
+
+Notification = namedtuple('Notification', 'event rules')
+
+
+def split_key(key):
+    from sentry.plugins import plugins  # XXX
+    plugin_slug, _, project_id = key.split(':', 2)
+    return plugins.get(plugin_slug), Project.objects.get(pk=project_id)
+
+
+def unsplit_key(plugin, project):
+    return '{plugin.slug}:p:{project.id}'.format(plugin=plugin, project=project)
+
+
+def strip_for_serialization(instance):
+    cls = type(instance)
+    return cls(**{field.attname: getattr(instance, field.attname) for field in cls._meta.fields})
+
+
+def event_to_record(event, rules):
+    if not rules:
+        logger.warning('Creating record for %r that does not contain any rules!', event)
+
+    return Record(
+        event.event_id,
+        Notification(strip_for_serialization(event), [rule.id for rule in rules]),
+        to_timestamp(event.datetime),
+    )
+
+
+def fetch_state(project, records):
+    # This reads a little strange, but remember that records are returned in
+    # reverse chronological order, and we query the database in chronological
+    # order.
+    # NOTE: This doesn't account for any issues that are filtered out later.
+    start = records[-1].datetime
+    end = records[0].datetime
+
+    groups = Group.objects.in_bulk(record.value.event.group_id for record in records)
+    return {
+        'project': project,
+        'groups': groups,
+        'rules': Rule.objects.in_bulk(itertools.chain.from_iterable(record.value.rules for record in records)),
+        'event_counts': tsdb.get_sums(tsdb.models.group, groups.keys(), start, end),
+        'user_counts': tsdb.get_distinct_counts_totals(tsdb.models.users_affected_by_group, groups.keys(), start, end),
+    }
+
+
+def attach_state(project, groups, rules, event_counts, user_counts):
```

Comment:
I hate that this is passed as `**kwargs` below but I'd rather have it doing some sort of parameter validation than just passing a random dictionary around, and alternative would be to either introduce a new type, or emulate the behavior of keyword argument unpacking, both of which seem a little excessive right now.